File: modules/wem-webapp/src/main/webapp/admin/common/js/content/ContentName.ts
module api.content {

    export class ContentName extends api.Name implements api.Equitable {

        public static UNNAMED_PREFIX: string = "__unnamed__";


        constructor(name: string) {
            super(name);
        }

        isUnnamed(): boolean {
            return false;
        }

        toUnnamed(): ContentUnnamed {
            api.util.assert(this instanceof ContentUnnamed, "this is not a ContentUnnamed");
            return <ContentUnnamed>this;
        }

        equals(o: api.Equitable): boolean {

            if (!(o instanceof ContentName)) {
                return false;
            }

            if (!super.equals(o)) {
                return false;
            }

            return true;
        }

        public static fromString(str: string) {

            api.util.assert(str != null, "name cannot be null");
            if (str.indexOf(ContentName.UNNAMED_PREFIX) == 0) {
                return new ContentUnnamed(str);
            }
            else {
                return new ContentName(str);
            }
        }

        public static ensureValidContentName(possibleInvalidName: string): string {
            if (!possibleInvalidName) {
                return "";
            }

            return Name.ensureValidName(possibleInvalidName);
        }
    }
}

File: modules/core/core-api/src/main/java/com/enonic/xp/content/ContentName.java
package com.enonic.xp.content;

import java.util.UUID;

import com.enonic.xp.annotation.PublicApi;
import com.enonic.xp.name.Name;

@PublicApi
public final class ContentName
    extends Name
{
    private static final String UNNAMED_PREFIX = "__unnamed__";

    private ContentName( final String name, boolean validate )
    {
        super( name, validate );
    }

    public boolean isUnnamed()
    {
        return this.value.startsWith( UNNAMED_PREFIX );
    }

    public boolean hasUniqueness()
    {
        return isUnnamed() && this.value.length() > UNNAMED_PREFIX.length();
    }

    public static ContentName from( final String name )
    {
        return new ContentName( name, true );
    }

    public static ContentName uniqueUnnamed()
    {
        return new ContentName( UNNAMED_PREFIX + UUID.randomUUID(), false );
    }
}
